src/ln/invoice/mod.rs

```diff
@@ -0,0 +1,187 @@
+
+use chrono::{DateTime, Utc, Duration};
+
+use regex::Regex;
+
+use secp256k1;
+use secp256k1::key::PublicKey;
+use secp256k1::{Signature, Secp256k1};
+
+mod parsers;
+
+/// An Invoice for a payment on the lightning network as defined in
+/// [BOLT #11](https://github.com/lightningnetwork/lightning-rfc/blob/master/11-payment-encoding.md#examples).
+
+#[derive(Eq, PartialEq, Debug)]
+pub struct Invoice {
+    /// The currency deferred from the 3rd and 4th character of the bech32 transaction
+    pub currency: Currency,
+
+    /// The amount to pay in pico-satoshis
+    pub amount: Option<u64>,
+
+    pub timestamp: DateTime<Utc>,
+
+    /// tagged fields of the payment request
+    pub tagged: Vec<TaggedField>,
+
+    pub signature: Signature,
+}
+
+#[derive(Eq, PartialEq, Debug)]
+pub enum Currency {
+    Bitcoin,
+    BitcoinTestnet,
+}
+
+#[derive(Eq, PartialEq, Debug)]
+pub enum TaggedField {
+    PaymentHash([u8; 32]),
+    Description(String),
+    PayeePubKey(PublicKey),
+    DescriptionHash([u8; 32]),
+    ExpiryTime(Duration),
+    MinFinalCltvExpiry(u64),
+    Fallback(Fallback),
+    Route {
+        pubkey: PublicKey,
+        short_channel_id: u64,
+        fee_base_msat: i32,
+        fee_proportional_millionths: i32,
+        cltv_expiry_delta: u16,
+    }
+}
+
+// TODO: better types instead onf byte arrays
+#[derive(Eq, PartialEq, Debug)]
+pub enum Fallback {
+    SegWitScript {
+        version: u8,
+        script: Vec<u8>,
+    },
+    PubKeyHash([u8; 20]),
+    ScriptHash([u8; 20]),
+}
+
+impl Invoice {
+    // TODO: maybe rewrite using nom
+    fn parse_hrp(hrp: &str) -> Result<(Currency, Option<u64>)> {
+        let re = Regex::new(r"^ln([^0-9]*)([0-9]*)([munp]?)$").unwrap();
+        let parts = match re.captures(&hrp) {
+            Some(capture_group) => capture_group,
+            None => return Err(ErrorKind::MalformedHRP.into())
+        };
+
+        let currency = parts[0].parse::<Currency>()?;
+
+        let amount = if !parts[1].is_empty() {
+            Some(parts[1].parse::<u64>()?)
+        } else {
+            None
+        };
+
+        /// `get_multiplier(x)` will only return `None` if `x` is not "m", "u", "n" or "p", which
+        /// due to the above regex ensures that `get_multiplier(x)` iif `x == ""`, so it's ok to
+        /// convert a none to 1BTC aka 10^12pBTC.
+        let multiplier = parts[2].chars().next().and_then(|suffix| {
+            get_multiplier(&suffix)
+        }).unwrap_or(1_000_000_000_000);
+
+        Ok((currency, amount.map(|amount| amount * multiplier)))
+    }
+}
+
+impl FromStr for Invoice {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self> {
+        let Bech32 {hrp, data} = s.parse()?;
+
+        let (currency, amount) = Invoice::parse_hrp(&hrp)?;
+
+        Ok(Invoice {
+            currency,
+            amount,
+            timestamp: Utc::now(),
+            tagged: vec![],
+            signature: Signature::from_der(&Secp256k1::new(), &[0; 65])?,
```

Comment:
Shouldn't this be from_compact not from_der (I believe this is busted elsewhere in the codebase right now...)